docs/technical/application-structure.md

```diff
@@ -0,0 +1,131 @@
+# Application Structure
+
+This document is a work in progress but covers where things can be found and
+explains how things are organized in the GitHub Desktop source.
+
+**Note:** this is an aspirational document and to be used as a starting point
+for discussions, and doesn't currently reflect reality.
+
+## `app/src`
+
+I'm going to walk through these in the order in what might look like a strange
+order, but hopefully it makes sense by the end.
+
+## Shared Modules
+
+These folders should contain modules that can be shared across any of the
+Webpack bundles that Desktop generates.
+
+ - `app/src/models` - contains the shapes used in the codebase to represent
+common objects. They should be immutable and plain.
+
+ - `app/src/lib` - contains functions that do not depend on executing in a
+ specific environment.
+
+In some cases, such as the main and renderer processes using IPC (inter-process
+communication), we want to ensure the shapes of objects being sent are
+consistent. For logic or features associated with a specific bundle and not
+intended to be shared across bundles, these should follow the same pattern and
+live within a `models` or `lib` folder within the given bundle directory.
+
+## Application Bundles
+
+I'm referring to these folders as "bundles" because our webpack config will
+transpile a specific file in each of these directories to generate what it
+needs to package and run the application.
+
+### Main Process - `app/src/main`
+
+Modules and logic to be bundled for the main process, which is the entry point
+for the user to launch Desktop.
+
+### User Interface - `app/src/renderer`
+
+Modules and logic to be bundled for the renderer process, which displays the
+user interface and handles most of the data management in Desktop.
+
+As the largest part of the current codebase, I've sketched out how the folders
+within `app/src/renderer` may be organized:
+
+ - app
+    - src
+       - renderer
+          - components
+            ... 
+            - dialogs
+            - primitives
+            - text
+            ...
+          - lib
+             - git
+          - models
+          - stores
+          - views
+
+There's a lot of opinions out there about how to structure React projects, but
+for the moment I wanted to focus on addressing these problems:
+
+ - better organization of our React components
+ - better organization of modules required by the renderer process
+ - clarification of modules that can be shared between parts of the app, and
+   things that make sense to keep specific to one part of the app
+ - reflect our current usage patterns
+    - Git operations performed in the renderer
+    - stores created and managed in the renderer
+
+What lives in each of these folders:
+
+ - `components` - contains the React components used in our application. I don't
```

Comment:
I've avoided going too deep on this because I didn't want to introduce too much churn in the codebase. Once we've gotten things more organized broadly, we can deal with this specific area.

There's lots of [examples out there](https://github.com/markerikson/react-redux-links/blob/master/project-structure.md) - including [my favourite](http://react-file-structure.surge.sh/), but I wanted to find one that felt natural for our use case, which I think we're still figuring out, so I'm happy to punt on that for now.

Reply:
sounds great!